File: server/src/mealog/pipeline/runner.py
"""Pipeline orchestration and the ablation configs.

Each config turns one idea on. That is what makes the scorecard an argument
rather than a number: V0 is what a single prompt gets you, and every row after
it has to pay for itself.
"""
from __future__ import annotations

from dataclasses import dataclass

from mealog import obs
from mealog.domain.models import MealLog, Nutrients, ResolvedItem
from mealog.locales.loader import LocalePack, load
from mealog.pipeline import normalize, nutrition, portion, retrieval
from mealog.pipeline.confidence import route
from mealog.pipeline.ports import VisionInput, VisionPort
from mealog.pipeline.resolve import resolve


@dataclass(frozen=True)
class Config:
    name: str
    description: str
    grounded: bool = True       # resolve against catalogue + compute nutrition
    locale_rules: bool = True   # apply locale text/unit normalization
    gating: bool = True         # confidence routing + abstention


CONFIGS: dict[str, Config] = {
    "V0": Config("V0", "single-prompt VLM, model reports calories directly",
                 grounded=False, locale_rules=False, gating=False),
    "V1": Config("V1", "+ closed-set resolution, nutrition computed from catalogue",
                 grounded=True, locale_rules=False, gating=False),
    "V2": Config("V2", "+ locale text and unit normalization",
                 grounded=True, locale_rules=True, gating=False),
    "V3": Config("V3", "+ confidence gating and abstention",
                 grounded=True, locale_rules=True, gating=True),
}


def _reconcile_resolved(items: list[ResolvedItem]) -> list[ResolvedItem]:
    """Collapse repeated grounded foods while preserving every abstention.

    ``ABSTAIN`` is a sentinel rather than a food, so different unmatched
    queries stay as separate questions. A known count can be added; one
    unknown contribution makes the merged count unknown.
    """
    by_food: dict[str, ResolvedItem] = {}
    reconciled: list[ResolvedItem] = []

    for item in items:
        if item.abstained:
            reconciled.append(item)
            continue

        existing = by_food.get(item.food_id)
        if existing is None:
            by_food[item.food_id] = item
            reconciled.append(item)
            continue

        existing.quantity = (
            None
            if existing.quantity is None or item.quantity is None
            else existing.quantity + item.quantity
        )
        existing.unit = existing.unit if existing.unit == item.unit else None
        existing.count_origin = (
            existing.count_origin
            if existing.count_origin == item.count_origin
            else "vision"
            if "vision" in {existing.count_origin, item.count_origin}
            else existing.count_origin or item.count_origin
        )
        existing.confidence = min(existing.confidence, item.confidence)

    return reconciled


def run(vision: VisionPort, input_ref: VisionInput | str, locale: str, config: Config,
        idempotency_key: str, text: str | None = None) -> MealLog:
    if isinstance(input_ref, str):
        # TODO(#3): Remove VisionInput | str compatibility once live callers pass
        # image-backed VisionInput objects.
        input_ref = VisionInput(sample_id=input_ref, text=text)
    elif text is not None:
        raise TypeError("text must be part of VisionInput")

    obs.new_request_id()
    pack: LocalePack = load(locale)
    log = MealLog(idempotency_key=idempotency_key, locale=locale, config=config.name)

    with obs.stage("perception", provider=vision.name, sample=input_ref.log_reference):
        perceived = vision.perceive(input_ref)

    if not config.grounded:
        # Baseline: trust whatever the model said. No catalogue, no arithmetic
        # of our own. This row exists to be beaten.
        log.items = [
            ResolvedItem(query=p.surface_form, food_id=f"ungrounded:{p.surface_form}",
                         confidence=p.confidence,
                         nutrients=Nutrients(kcal=p.ungrounded_kcal or 0.0))
            for p in perceived
        ]
        log.totals = sum((i.nutrients for i in log.items), Nutrients()).rounded()
        log.action = "auto_accept"
        return log

    with obs.stage("normalize", locale=locale, rules=config.locale_rules):
        normalized = normalize.normalize(perceived, pack, apply_rules=config.locale_rules)

    resolved: list[ResolvedItem] = []
    for item in normalized:
        with obs.stage("retrieval", query=item.query):
            candidates = retrieval.search(item.query, pack)
        r = resolve(item.query, candidates, allow_abstain=config.gating)
        r.quantity = item.quantity
        r.unit = item.unit
        r.count_origin = item.count_origin
        resolved.append(r)

    reconciled = _reconcile_resolved(resolved)
    for item in reconciled:
        if item.abstained:
            continue

        food = pack.foods[item.food_id]
        # A missing count is not an implicit one. Keep the item on the
        # catalogue default path even when an uncounted hint names a unit.
        estimate = portion.estimate(
            food,
            item.quantity,
            item.unit if item.quantity is not None else None,
            pack,
            item.count_origin,
        )
        item.grams, item.grams_p10, item.grams_p90 = estimate
        item.portion_source = estimate.source
        item.portion_provenance = estimate.provenance
        item.nutrients = nutrition.scale_per_100g(food.per_100g, item.grams).rounded()

    log.items = reconciled
    log.totals = sum((i.nutrients for i in reconciled if not i.abstained), Nutrients()).rounded()
    if config.gating:
        log = route(log)
    else:
        log.action = "auto_accept"

    obs.event("meal_logged", config=config.name, locale=locale, action=log.action,
              items=len(reconciled), abstained=sum(i.abstained for i in reconciled),
              kcal=log.totals.kcal)
    return log


def make_vision(provider: str, api_key: str | None = None) -> VisionPort:
    if provider == "fixture":
        from mealog.adapters.vision_fixture import FixtureVision
        return FixtureVision()
    if provider == "gemini":
        from mealog.adapters.vision_gemini import GeminiVision
        return GeminiVision(api_key or "")
    raise ValueError(f"unknown vision provider: {provider}")

File: server/src/pipeline/runner.ts
/**
 * Pipeline orchestration and ablation configurations.
 *
 * This is a framework-free port of `server/src/mealog/pipeline/runner.py`.
 * The only external boundary is VisionPort: adapters are supplied by the
 * caller, while every grounded stage remains a pure pipeline module.
 */

import {
  ABSTAIN,
  addNutrients,
  makeMealLog,
  makeNutrients,
  makeResolvedItem,
  roundedNutrients,
  type MealLog,
  type ResolvedItem,
} from '../domain/models';
import { load } from '../locales/loader';
import { normalize } from './normalize';
import { estimate } from './portion';
import { createRetrieval } from './retrieval/index';
import { resolve } from './resolve';
import { captureMediumQuestion, route } from './confidence';
import { addClarifications } from './clarification';
import { scalePer100g } from './nutrition';
import { VisionInput, type VisionPort } from './ports';

export interface Config {
  readonly name: string;
  readonly description: string;
  readonly grounded: boolean;
  readonly locale_rules: boolean;
  readonly gating: boolean;
}

export const CONFIGS: Readonly<Record<string, Config>> = {
  V0: {
    name: 'V0',
    description: 'single-prompt VLM, model reports calories directly',
    grounded: false,
    locale_rules: false,
    gating: false,
  },
  V1: {
    name: 'V1',
    description: '+ closed-set resolution, nutrition computed from catalogue',
    grounded: true,
    locale_rules: false,
    gating: false,
  },
  V2: {
    name: 'V2',
    description: '+ locale text and unit normalization',
    grounded: true,
    locale_rules: true,
    gating: false,
  },
  V3: {
    name: 'V3',
    description: '+ confidence gating and abstention',
    grounded: true,
    locale_rules: true,
    gating: true,
  },
};

const retrieval = createRetrieval();

/**
 * Collapse repeated grounded observations without ever collapsing abstentions.
 * ABSTAIN is a sentinel, not a food, so two different unmatched queries must
 * remain visible as two questions. A known count can be added; one unknown
 * contribution makes the merged count unknown.
 */
function reconcileResolved(items: ResolvedItem[]): ResolvedItem[] {
  const byFood = new Map<string, ResolvedItem>();
  const reconciled: ResolvedItem[] = [];

  for (const item of items) {
    if (item.food_id === ABSTAIN) {
      reconciled.push(item);
      continue;
    }

    const existing = byFood.get(item.food_id);
    if (existing === undefined) {
      byFood.set(item.food_id, item);
      reconciled.push(item);
      continue;
    }

    existing.quantity = existing.quantity === null || item.quantity === null
      ? null
      : existing.quantity + item.quantity;
    existing.unit = existing.unit === item.unit ? existing.unit : null;
    existing.count_origin = existing.count_origin === item.count_origin
      ? existing.count_origin
      : existing.count_origin === 'vision' || item.count_origin === 'vision'
        ? 'vision'
        : existing.count_origin ?? item.count_origin;
    if (existing.capture_medium === 'real_plate' && item.capture_medium !== 'real_plate') {
      existing.capture_medium = item.capture_medium;
    }
    existing.confidence = Math.min(existing.confidence, item.confidence);
  }

  return reconciled;
}

function ungroundedLog(
  perceived: Awaited<ReturnType<VisionPort['perceive']>>['observations'],
  degraded: boolean,
  idempotencyKey: string,
  locale: string,
  config: Config,
): MealLog {
  const items = perceived.map((item) => makeResolvedItem({
    query: item.surface_form,
    food_id: `ungrounded:${item.surface_form}`,
    confidence: item.confidence,
    capture_medium: item.capture_medium,
    nutrients: makeNutrients({ kcal: item.ungrounded_kcal || 0.0 }),
  }));

  const totals = items.reduce(
    (acc, item) => addNutrients(acc, item.nutrients),
    makeNutrients(),
  );
  const mediumFlag = items.find((item) => item.capture_medium !== 'real_plate');
  return makeMealLog({
    idempotency_key: idempotencyKey,
    locale,
    config: config.name,
    items,
    totals: roundedNutrients(totals),
    action: degraded ? 'review' : mediumFlag === undefined ? 'auto_accept' : 'ask',
    question: mediumFlag === undefined ? null : captureMediumQuestion(mediumFlag),
    degraded,
  });
}

/**
 * Run one meal through perception, normalization, retrieval, resolution,
 * portion, nutrition, and (for V3) confidence routing.
 *
 * `inputRef: string` preserves the Python runner's fixture compatibility path;
 * live callers should pass a VisionInput. A provided VisionPort may be sync or
 * async, which keeps handwritten test stubs as small as the production port.
 */
export async function run(
  vision: VisionPort,
  inputRef: VisionInput | string,
  locale: string,
  config: Config,
  idempotencyKey: string,
  text: string | null = null,
): Promise<MealLog> {
  let input: VisionInput;
  if (typeof inputRef === 'string') {
    input = new VisionInput({ sampleId: inputRef, text });
  } else {
    if (text !== null) {
      throw new TypeError('text must be part of VisionInput');
    }
    input = inputRef;
  }

  const pack = load(locale);
  const perception = await vision.perceive(input);
  const perceived = perception.observations;
  const degraded = perception.degraded;

  if (!config.grounded) {
    return ungroundedLog(perceived, degraded, idempotencyKey, locale, config);
  }

  const normalized = normalize(
    perceived,
    pack,
    config.locale_rules,
  );
  const resolved: ResolvedItem[] = [];

  for (const item of normalized) {
    const candidates = retrieval.search(item.query, pack);
    const result = resolve(item.query, candidates, config.gating);
    result.quantity = item.quantity;
    result.unit = item.unit;
    result.count_origin = item.count_origin;
    result.capture_medium = item.original.capture_medium;

    // Keep ABSTAIN as an item. In particular, do not run it through portion or
    // nutrition: a missing catalogue food is not a zero-calorie food.
    resolved.push(result);
  }

  const reconciled = reconcileResolved(resolved);
  for (const result of reconciled) {
    if (result.food_id === ABSTAIN) {
      continue;
    }

    const food = pack.foods[result.food_id];
    // A missing count is not an implicit one. Keep the item on the catalogue
    // default path even when an uncounted hint happened to name a unit.
    const portion = estimate(
      food,
      result.quantity,
      result.quantity === null ? null : result.unit,
      pack,
      undefined,
      result.count_origin,
    );
    result.grams = portion.grams;
    result.grams_p10 = portion.p10;
    result.grams_p90 = portion.p90;
    result.portion_source = portion.source;
    result.portion_provenance = portion.provenance;
    result.nutrients = roundedNutrients(scalePer100g(food.per_100g, result.grams));
  }

  const totals = reconciled
    .filter((item) => item.food_id !== ABSTAIN)
    .reduce(
      (acc, item) => addNutrients(acc, item.nutrients),
      makeNutrients(),
    );
  let log = makeMealLog({
    idempotency_key: idempotencyKey,
    locale,
    config: config.name,
    items: reconciled,
    totals: roundedNutrients(totals),
    degraded,
  });

  if (degraded) {
    // Provider fallback is never a first-class accepted answer, regardless of
    // identity or portion confidence. The status came from this request's
    // perception envelope, not from mutable adapter state.
    log.action = 'review';
  } else if (config.gating) {
    log = route(log);
  } else {
    const mediumFlag = reconciled.find((item) => item.capture_medium !== 'real_plate');
    if (mediumFlag !== undefined) {
      log.action = 'ask';
      log.question = captureMediumQuestion(mediumFlag);
    } else {
      log.action = 'auto_accept';
    }
  }
  return addClarifications(log, pack);
}
